### ISSUE TITLE
Volume can't be mounted if already exists in volume attachments list and was not removed from it when pod was stopped

### ISSUE THREAD

**Issue by dimm0** [rook/rook#1214]
After a volume has been created and mounted to a pod, and after a pod node died or something else prevented volume from detaching clearly, the record remains in volume attachments list. The record prevents the volume from being attached again to a new pod from the same group.

**Comment by bassam** [comment]
@dimm0 what cases lead to a pod diring and the volume not detaching cleanly? Would be great to see steps for reproducing this.

**Comment by dimm0** [comment]
I'll update the ticket once I can reproduce this

**Comment by jbw976** [comment]
I just sat down with @dimm0 while he patiently worked with me to walk through his repro scenario.  It looks like we can reproduce this by attempting to attach a massive volume to a pod, e.g. 5 TB.  What appears to happen is that `rbd map` will take a long time (@dimm0 said this was normal for the in-tree rbd plugin also), but our agent will kill rbd before it's done.  Then when the user deletes the pod to try to schedule it somewhere else, k8s does NOT go through the detach flow, so we will never clean up the 1st volume attachment we were using for locking.

Basically, if we give up on `rbd` too early, and the pod moves, the detach flow will not occur because k8s didn't think the volume was finished attaching.

Perhaps k8s doesn't think the attachment is done until the end, while we are doing our locking by creating a volume attachment CRD at the beginning.

We should be able to repro this now and figure out a fix.

**Comment by bassam** [comment]
does it make sense for us to remove the CRD instance on failures like this? Or is there a different way to trigger K8S to detach? Also it sounds like we need to adjust the timeouts.

I think we should consider back porting this 0.6

**Comment by kokhang** [comment]
I think we should do both.
1- Increase rbd timeout. Currently is 1min. What is a good timeout that will handle rbd command to map a massive volume? 3mins?
2- Delete VolumeAttachment CRD record if attachment failed.

**Comment by paha** [comment]
Can the timeout be dynamic, calculating timeout based on the size etc. So that we wouldn't hit the issue later on even bigger volumes?

**Comment by dimm0** [comment]
I just hit this recently: http://lists.ceph.com/pipermail/ceph-users-ceph.com/2017-October/022000.html
Looks like the issue we're having here... But not sure how I can test whether this is the solution

**Comment by jbw976** [comment]
@dimm0 i think that issue is actually different (even though it's useful information) because it is referring to the formatting of the disk using `mkfs`.  we are seeing a very long hang on the `rbd map` operation, before we even get to formatting it.

@bassam had asked about how the in-tree rbd plugin behaves.  it looks like when it executes `rbd map`, it does not have any timeout logic at all, so an exceptionally long `map` call would just wait: https://github.com/kubernetes/kubernetes/blob/a7e76475ef784a2e5f626ed36c9e4c89331e5724/pkg/volume/rbd/rbd_util.go#L286

when @kokhang is back online, we should focus on:
1. understanding why `rbd map` can take many minutes for a large volume, if that is expected, and how to know the operation is making progress (as opposed to just being hung, which we have seen sometimes even for small volumes, which led us to introduce the timeout in the first place).
1. error handling and CRD clean up.  we want to harden our approach and not leak any volume attach CRDS that require manual user intervention.

**Comment by kokhang** [comment]
@jbw976 @bassam i believe Rook and the on-tree RBD would result on the same behavior. Both uses the RBD command to map the image. Here is how rook does it: https://github.com/rook/rook/blob/4509ecab4a9b88d252c7ebbb0c4bdba03bd4735e/pkg/ceph/client/image.go#L109. And this is where it hangs.

**Comment by kokhang** [comment]
@jbw976 @dimm0 I have started working on a cleanup but i am started asking myself if it's really needed. Perhaps i am not quite sure how to repro the issue you are experiencing.

> Basically, if we give up on rbd too early, and the pod moves, the detach flow will not occur because k8s didn't think the volume was finished attaching.

Correct that this will leave an orphaned VolumeAttachment record in the CRD.
@jbw976 but when you says "the pod moves", does that mean that it is deleted and a new pod (with new ID) is created somewhere else? If so, the current logic takes care of that. It determines that the volumeattachment is orphaned and will update it to allow the attachment to happen on the new pod.

@dimm0 what type of resource did you use to reproduce this? Was it a statefulset, deployment or just pod?

**Comment by dimm0** [comment]
@kokhang I first saw this on statefulset, but mostly because those take longer to init and are easier to break. I think I also saw this on deployments.
Pretty much any error mode, when pod soesn't fully start, will cause it. That's why it's hard to reproduce - you need another error to get this error :)

**Comment by kokhang** [comment]
i am able to generate the first error and i can see that the VolumeAttachment are not deleted. But that is already handled by the logic if the original pod is deleted.

**Comment by dimm0** [comment]
Is that a recent change? Maybe it's really fixed in a recent rook release, if you're clearing an orphaned attachments

**Comment by kokhang** [comment]
i dont think there was a recent change. This issue is handled here https://github.com/rook/rook/blob/master/pkg/agent/flexvolume/controller.go#L104.

So i created a statefulset and mocked the `rbd` command to hang in order to reproduce the error. This is the log im getting:

```2017-11-30 00:40:07.735970 I | rook-flexdriver: calling agent to attach volume replicapool/pvc-5ec2bb0b-d564-11e7-abea-08002751a1e0
2017-11-30 00:40:07.738969 I | ceph-volumeattacher: attaching volume replicapool/pvc-5ec2bb0b-d564-11e7-abea-08002751a1e0 cluster rook
2017-11-30 00:40:07.742133 I | cephmon: parsing mon endpoints: rook-ceph-mon0=10.0.0.52:6790,rook-ceph-mon1=10.0.0.174:6790,rook-ceph-mon2=10.0.0.130:6790
2017-11-30 00:40:07.742216 I | op-mon: loaded: maxMonID=2, mons=map[rook-ceph-mon0:0xc4204a18a0 rook-ceph-mon1:0xc4204a1900 rook-ceph-mon2:0xc4204a1940], mapping=&{Node:map[rook-ceph-mon0:0xc4204a19a0 rook-ceph-mon1:0xc4204a1a00 rook-ceph-mon2:0xc4204a1a80] Port:map[]}
2017-11-30 00:40:07.742356 I | exec: Running command: rbd map replicapool/pvc-5ec2bb0b-d564-11e7-abea-08002751a1e0 --id admin --cluster=rook --keyring=/tmp/rook.keyring871401260 -m 10.0.0.52:6790,10.0.0.174:6790,10.0.0.130:6790 --conf=/dev/null
2017-11-30 00:41:07.743538 I | exec: Timeout waiting for process rbd to return. Sending interrupt signal to the process
2017-11-30 00:42:07.745732 I | exec: Timeout waiting for process rbd to return after interrupt signal was sent. Sending kill signal to the process
2017-11-30 00:42:07.746873 E | rook-flexdriver: Attach volume replicapool/pvc-5ec2bb0b-d564-11e7-abea-08002751a1e0 failed: failed to attach volume replicapool/pvc-5ec2bb0b-d564-11e7-abea-08002751a1e0: failed to map image replicapool/pvc-5ec2bb0b-d564-11e7-abea-08002751a1e0 cluster rook. failed to map image replicapool/pvc-5ec2bb0b-d564-11e7-abea-08002751a1e0: Failed to complete rbd: Timeout waiting for the command rbd to return. output:
2017-11-30 00:42:15.980517 I | rook-flexdriver: calling agent to attach volume replicapool/pvc-5ec2bb0b-d564-11e7-abea-08002751a1e0
2017-11-30 00:42:15.985872 I | ceph-volumeattacher: attaching volume replicapool/pvc-5ec2bb0b-d564-11e7-abea-08002751a1e0 cluster rook
2017-11-30 00:42:15.990473 I | cephmon: parsing mon endpoints: rook-ceph-mon0=10.0.0.52:6790,rook-ceph-mon1=10.0.0.174:6790,rook-ceph-mon2=10.0.0.130:6790
2017-11-30 00:42:15.990524 I | op-mon: loaded: maxMonID=2, mons=map[rook-ceph-mon1:0xc4204a0760 rook-ceph-mon2:0xc4204a07c0 rook-ceph-mon0:0xc4204a0720], mapping=&{Node:map[rook-ceph-mon2:0xc4204a08e0 rook-ceph-mon0:0xc4204a0840 rook-ceph-mon1:0xc4204a08a0] Port:map[]}
2017-11-30 00:42:15.990673 I | exec: Running command: rbd map replicapool/pvc-5ec2bb0b-d564-11e7-abea-08002751a1e0 --id admin --cluster=rook --keyring=/tmp/rook.keyring905771675 -m 10.0.0.52:6790,10.0.0.174:6790,10.0.0.130:6790 --conf=/dev/null```

This is running as expected. It is calling the `rbd` to attach. It is hanging. The rbd processed is killed and it retries again.

@dimm0 Could you share your manifest yaml to see if im missing something that could reproduce the issue?

**Comment by dimm0** [comment]
Ok, I've got the working one :)
Here's the yaml: https://github.com/dimm0/prp_k8s_config/blob/master/volume-example.yaml
Replace "storage: 20Gi" with some large number, I used "50Ti"
```
Dmitrys-MacBook-Pro-2:prp_k8s_config dimm$ kubectl create -f volume-example.yaml
persistentvolumeclaim "examplevol" created
pod "vol-pod" created
```
Then wait until it fails to mount:
<img width="985" alt="screen shot 2017-11-29 at 9 17 38 pm" src="https://user-images.githubusercontent.com/1425007/33415410-34420bba-d54b-11e7-979b-b94867a58d65.png">

Delete pod:

```
Dmitrys-MacBook-Pro-2:prp_k8s_config dimm$ kubectl delete pod vol-pod
pod "vol-pod" deleted
```
<img width="986" alt="screen shot 2017-11-29 at 9 18 15 pm" src="https://user-images.githubusercontent.com/1425007/33415414-3b03d2b2-d54b-11e7-8a56-5f86699326cd.png">

Recreate pod:

```
Dmitrys-MacBook-Pro-2:prp_k8s_config dimm$ kubectl create -f volume-example.yaml
pod "vol-pod" created
Error from server (AlreadyExists): error when creating "volume-example.yaml": persistentvolumeclaims "examplevol" already exists
```

(Persistent volume still exists - same happens when using a deployment and deleting a pod in it.)

<img width="991" alt="screen shot 2017-11-29 at 9 19 01 pm" src="https://user-images.githubusercontent.com/1425007/33415427-57e80f7e-d54b-11e7-8986-220b5884472c.png">

Here it is. Can't mount until deleted from attached volumes list manually.

**Comment by dimm0** [comment]
By the way, after deleting both pod and its claim, the volume is still hanging there. As I remember it will never be deleted...

```
Dmitrys-MacBook-Pro-2:prp_k8s_config dimm$ kubectl delete -f volume-example.yaml
persistentvolumeclaim "examplevol" deleted
pod "vol-pod" deleted
```

<img width="996" alt="screen shot 2017-11-29 at 9 34 22 pm" src="https://user-images.githubusercontent.com/1425007/33415709-3cc45890-d54d-11e7-9708-2c530465a787.png">

```
Dmitrys-MacBook-Pro-2:prp_k8s_config dimm$ kubectl get volumeattachment --all-namespaces -o json | grep pvc-3c185b73-d58d-11e7-a68f-0cc47a6a1e1e
                    "mountDir": "/var/lib/kubelet/pods/3c200623-d58d-11e7-a68f-0cc47a6a1e1e/volumes/rook.io~rook/pvc-3c185b73-d58d-11e7-a68f-0cc47a6a1e1e",
                "name": "pvc-3c185b73-d58d-11e7-a68f-0cc47a6a1e1e",
                "selfLink": "/apis/rook.io/v1alpha1/namespaces/rook-system/volumeattachments/pvc-3c185b73-d58d-11e7-a68f-0cc47a6a1e1e",
```

**Comment by kokhang** [comment]
@dimm0 ahh you are using the Pod resource. Since the pod name remains the same but the ID differs, the fencing logic doesnt know whether the pods are the same. Ok i have a fix for this. Thanks!

**Comment by dimm0** [comment]
As I said, same applies to deleting a pod inside a deployment

**Comment by kokhang** [comment]
I will use this issue to track the VolumeAttachment leak issue since it has a lot of the content. I have created #1281 to track the RBD mapping hang on large volume issue.
